Slow stop command

I'm noticing on one of our nodes that `stop` command waits for something and that is why stop can take a few seconds to complete. I'm not sure if this is connected but also when doing `ioc console myjail` cli hangs for a few seconds. I'm really struggling to find the cause. Please check stop/start calls and some timing. Also one thing to consider: there is no service to be started or stopped `service -e`
 show 
```
/etc/rc.d/cleanvar
/etc/rc.d/newsyslog
/etc/rc.d/syslogd
/etc/rc.d/virecover
/etc/rc.d/motd
```

```
# time ioc start ioc/abruptive-firefly-4GBLKB
[+] JailResolverConfig@ioc/abruptive-firefly-4GBLKB: OK [0.001s]
[+] JailDependantsStart@ioc/abruptive-firefly-4GBLKB: No dependant jails [0.0s]
[+] JailLaunch@ioc/abruptive-firefly-4GBLKB: OK [0.235s]
abruptive-firefly-4GBLKB running as JID 39
0.300u 0.218s 0:00.41 124.3%	22+167k 3486+16io 1pf+0w
# time ioc stop ioc/abruptive-firefly-4GBLKB
[+] JailDestroy@ioc/abruptive-firefly-4GBLKB: OK [5.91s]
abruptive-firefly-4GBLKB stopped
4.345u 1.850s 0:06.13 100.9%	6+167k 1656+5io 0pf+0w
# time ioc start ioc/abruptive-firefly-4GBLKB
[+] JailResolverConfig@ioc/abruptive-firefly-4GBLKB: OK [0.001s]
[+] JailDependantsStart@ioc/abruptive-firefly-4GBLKB: No dependant jails [0.0s]
[+] JailLaunch@ioc/abruptive-firefly-4GBLKB: OK [0.236s]
abruptive-firefly-4GBLKB running as JID 40
0.310u 0.209s 0:00.41 124.3%	28+172k 3486+16io 1pf+0w
# time ioc stop ioc/abruptive-firefly-4GBLKB
[+] JailDestroy@ioc/abruptive-firefly-4GBLKB: OK [5.854s]
abruptive-firefly-4GBLKB stopped
4.265u 1.873s 0:06.07 100.9%	6+168k 1656+5io 0pf+0w
# time ioc start ioc/abruptive-firefly-4GBLKB
[+] JailResolverConfig@ioc/abruptive-firefly-4GBLKB: OK [0.001s]
[+] JailDependantsStart@ioc/abruptive-firefly-4GBLKB: No dependant jails [0.0s]
[+] JailLaunch@ioc/abruptive-firefly-4GBLKB: OK [0.236s]
abruptive-firefly-4GBLKB running as JID 41
0.304u 0.216s 0:00.41 124.3%	22+169k 3486+16io 1pf+0w
 # time ioc stop ioc/abruptive-firefly-4GBLKB
[+] JailDestroy@ioc/abruptive-firefly-4GBLKB: OK [5.493s]
abruptive-firefly-4GBLKB stopped
4.099u 1.674s 0:05.71 100.8%	5+167k 1656+5io 0pf+0w
```
I've noticed *python3.6 being WCPU on 93% during this slowness. So it looks something with python after all.

> what are fixed JIDs?

@igalic when you start a jail with a given jid (e.g. `jail -c persist path=/rescue jid=123`).

> As long as you don't have fixed JIDs, you should be able to restart the jail while the old one is DYING, iirc the check for name collisions ignores dying jails.

The current issue is that the NullFS release mounts cannot be unmounted after jail removal (and awaiting it to leave the dying state). In consequence the ZFS datasets they are mounted on top cannot be deleted. This issue occurred after changing to the following:

1. Jail removal with libc.jail_remove() instead of forking `jail -r`
  a. awaiting until `libc.jail_get(..., JAIL_DYING)` does fail or return 0
2. Replacing `umount {path}` with `libc.unmount({path})`

https://cirrus-ci.com/task/5661599502172160 shows the following

```
tests/test_Jail.py::TestNullFSBasejail::test_can_be_started ERROR        [ 61%]
==================================== ERRORS ====================================
_________ ERROR at teardown of TestNullFSBasejail.test_can_be_started __________
tp = <class 'libzfs.ZFSException'>, value = None, tb = None
    def reraise(tp, value, tb=None):
        try:
            if value is None:
                value = tp()
            if value.__traceback__ is not tb:
                raise value.with_traceback(tb)
>           raise value
/usr/local/lib/python3.6/site-packages/six.py:693: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.6/site-packages/six.py:693: in reraise
    raise value
/usr/local/lib/python3.6/site-packages/six.py:693: in reraise
    raise value
tests/conftest.py:203: in new_jail
    new_jail.destroy()
tests/libioc/Jail.py:2350: in destroy
    return list(JailGenerator.destroy(self, force=force))
tests/libioc/Jail.py:1092: in destroy
    raise e
tests/libioc/Jail.py:1089: in destroy
    self.zfs.delete_dataset_recursive(self.dataset)
tests/libioc/ZFS.py:96: in delete_dataset_recursive
    self.delete_dataset_recursive(child)
tests/libioc/ZFS.py:101: in delete_dataset_recursive
    dataset.umount()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   libzfs.ZFSException: Device busy
libzfs.pyx:2547: ZFSException
--------------------------- Captured stdout teardown ---------------------------
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/bin unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/boot unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/lib unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/libexec unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/rescue unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/sbin unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
/.ioc-test-12.0-RELEASE/libioc-test/jails/new-jail-2935/root/.iocage unmount failed
DEBUG DEBUG
b'USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W NAME\n'
===================== 8 passed, 1 error in 136.83 seconds ======================
*** Error code 1
```

The first step is to jail_remove() and wait for its death.
https://github.com/bsdci/libioc/blob/64dc7e74a9317b579e5c6d957e884d7752636ac3/libioc/Jail.py#L1493-L1496

Then the following code runs to move a jails mountpoints.
https://github.com/bsdci/libioc/blob/64dc7e74a9317b579e5c6d957e884d7752636ac3/libioc/Jail.py#L2093-L2107

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -471,6 +471,9 @@ class JailGenerator(JailResource):
         ]
         exec_poststart: typing.List[str] = []
 
+        if self.host.ipfw_enabled is True:
+            exec_start += ["set +e", "service ipfw onestop", "set -e"]
+
         if self.config["vnet"]:
             _created, _start = self._start_vimage_network()
             exec_created += _created
@@ -1718,9 +1721,13 @@ class JailGenerator(JailResource):
             ]
         ))
 
-        _ipfw_enabled = self.host.ipfw_enabled
         self._write_hook_script("command", "\n".join(
-            (["set +e", "service ipfw onestop"] if _ipfw_enabled else []) + [
+            (
+                [
+                    "set +e",
+                    "/usr/sbin/service ipfw onestop"
+                ] if self.host.ipfw_enabled else []
+            ) + [
                 "set -e",
                 f". {self._relative_hook_script_dir}/start.sh",
                 jail_command,

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -1488,7 +1488,7 @@ class JailGenerator(JailResource):
         jid = self.jid
         try:
             libjail.dll.jail_remove(jid)
-            while libjail.is_jid_dying(jid) is True:
+            while self.running or (libjail.is_jid_dying(jid) is True):
                 # wait for death
                 continue
             self.__jid = None

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -54,8 +54,6 @@ import libioc.Config.Jail.Properties.ResourceLimit
 import libioc.ResourceSelector
 import libioc.Config.Jail.File.Fstab
 
-import jail as libjail
-
 
 class JailResource(
     libioc.LaunchableResource.LaunchableResource,
@@ -566,15 +564,17 @@ class JailGenerator(JailResource):
                 stdout, stderr, returncode = self._launch_persistent_jail(
                     passthru=passthru
                 )
-                self.query_jid()
             else:
                 stdout, stderr, returncode = self._launch_single_command_jail(
                     single_command,
                     passthru=passthru
                 )
                 scope = jailLaunchEvent.scope
+                jid = self.jid
                 yield from self.__destroy_jail(scope)
                 yield from self.__run_hook("poststop", False, scope)
+                if jid is not None:
+                    yield from self.__stop_network(jid, False, scope)
                 yield from self.__teardown_mounts(False, scope)
                 yield from self.__clear_resource_limits(False, scope)
 
@@ -930,11 +930,14 @@ class JailGenerator(JailResource):
         jailStopEvent = events.JailStop(self, scope=event_scope)
 
         yield jailStopEvent.begin()
+        jid = self.jid
 
         yield from self.__run_hook("prestop", force, jailStopEvent.scope)
         yield from self.__run_hook("stop", force, jailStopEvent.scope)
         yield from self.__destroy_jail(jailStopEvent.scope)
         yield from self.__run_hook("poststop", force, jailStopEvent.scope)
+        if jid is not None:
+            yield from self.__stop_network(jid, force, jailStopEvent.scope)
         yield from self.__teardown_mounts(force, jailStopEvent.scope)
         yield from self.__clear_resource_limits(force, jailStopEvent.scope)
 
@@ -1646,6 +1649,7 @@ class JailGenerator(JailResource):
             )
             return stdout, stderr, returncode
 
+        self.query_jid()
         self.logger.verbose(
             f"Jail '{self.humanreadable_name}' started with JID {self.jid}"
         )
@@ -1729,6 +1733,7 @@ class JailGenerator(JailResource):
         )
 
         if returncode > 0:
+            self.query_jid()
             message = f"Jail {self.humanreadable_name} command failed."
         else:
             message = f"Jail {self.humanreadable_name} command finished."
@@ -1832,18 +1837,35 @@ class JailGenerator(JailResource):
 
         return created, start
 
-    def _stop_network(self) -> typing.List[str]:
-        if self.config["vnet"]:
-            return self._stop_vimage_network()
-        else:
-            return self._stop_non_vimage_network()
+    def __stop_network(
+        self,
+        jid: int,
+        force: bool,
+        event_scope: 'libioc.events.Scope'
+    ) -> typing.Generator['libioc.events.JailHook', None, None]:
+        teardownJailNetworkEvent = libioc.events.TeardownJailNetwork(
+            jail=self,
+            scope=event_scope
+        )
+        yield teardownJailNetworkEvent.begin()
+        try:
+            if self.config["vnet"] is True:
+                self._stop_vimage_network(jid)
+            else:
+                self._stop_non_vimage_network()
+        except Exception:
+            yield teardownJailNetworkEvent.fail()
+            if force is False:
+                raise
+            else:
+                return
+        yield teardownJailNetworkEvent.end()
 
-    def _stop_non_vimage_network(self) -> typing.List[str]:
-        commands: typing.List[str] = []
+    def _stop_non_vimage_network(self) -> None:
         for protocol in (4, 6,):
             config_value = self.config[f"ip{protocol}_addr"]
             if config_value is None:
-                return commands
+                continue
             for nic, addresses in config_value.items():
                 if addresses is None:
                     continue
@@ -1852,16 +1874,13 @@ class JailGenerator(JailResource):
                         # skip DHCP and ACCEPT_RTADV
                         continue
                     inet = "inet" if (protocol == 4) else "inet6"
-                    commands.append(
-                        f"/sbin/ifconfig {nic} {inet} {address} remove"
+                    libioc.helpers.exec(
+                        ["/sbin/ifconfig", nic, inet, str(address), "remove"]
                     )
-        return commands
 
-    def _stop_vimage_network(self) -> typing.List[str]:
-        commands: typing.List[str] = []
+    def _stop_vimage_network(self, jid: int) -> None:
         for network in self.networks:
-            commands += network.teardown()
-        return commands
+            network.teardown(jid)
 
     def _configure_localhost_commands(self) -> typing.List[str]:
         return ["/sbin/ifconfig lo0 localhost"]
@@ -2082,6 +2101,7 @@ class JailGenerator(JailResource):
                     reason=f"failed to unmount mountpoints of jail {name}",
                     logger=self.logger
                 )
+            return
 
         yield teardownJailMountsEvent.end()
 

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -1067,14 +1067,12 @@ class JailGenerator(JailResource):
 
         if _stop_jail is True:
             try:
-                stop_events = JailGenerator.stop(
+                yield from JailGenerator.stop(
                     self,
                     force=True,
                     event_scope=event_scope,
                     log_errors=(force_stop is False)
                 )
-                for event in stop_events:
-                    yield event
             except libioc.errors.JailDestructionFailed:
                 pass
 

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -24,6 +24,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 """iocage Jail module."""
 import typing
+import ctypes
 import os
 import random
 import shlex
@@ -2090,10 +2091,14 @@ class JailGenerator(JailResource):
         ))
 
         error = False
+
+        umount_flags = ctypes.c_ulonglong(0x80000)
+
         for mountpoint in (fstab_destinations + system_mountpoints):
             if os.path.ismount(mountpoint) is False:
                 continue
-            if libjail.dll.unmount(mountpoint.encode("utf-8")) == 0:
+            _mountpoint = mountpoint.encode("utf-8")
+            if libjail.dll.unmount(_mountpoint, umount_flags) == 0:
                 self.logger.verbose(f"{mountpoint} successfully unmounted")
             else:
                 error = True

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -561,7 +561,6 @@ class JailGenerator(JailResource):
             self._save_autoconfig()
 
         try:
-            self._prepare_stop()
             if single_command is None:
                 stdout, stderr, returncode = self._launch_persistent_jail(
                     passthru=passthru
@@ -847,56 +846,55 @@ class JailGenerator(JailResource):
         if os.path.isdir(realpath) is False:
             os.makedirs(realpath, 0o755)
 
-    def _prepare_stop(self) -> None:
-        self._ensure_script_dir()
+    def __run_hook(
+        self,
+        hook: str,
+        force: bool,
+        event_scope: 'libioc.events.Scope'
+    ) -> typing.Generator['libioc.events.JailHook', None, None]:
 
-        exec_prestop = []
-        exec_stop = []
-        exec_poststop = self._teardown_mounts() + self._clear_resource_limits()
+        exec_func = libioc.helpers.exec
+        exec_args = dict(logger=self.logger, env=self.env)
+        if hook == "prestop":
+            Event = libioc.events.JailHookPrestop
+        elif hook == "stop":
+            Event = libioc.events.JailHookStop
+            exec_func = self.exec
+            exec_args = dict()
+        elif hook == "poststop":
+            Event = libioc.events.JailHookPoststop
+        else:
+            raise NotImplementedError(f"jail hook {hook} is unsupported")
 
-        # ToDo: self.config.get("exec_prestop", "")
-        if self.config["exec_prestop"] is not None:
-            exec_prestop.append(self.config["exec_prestop"])
-        if self.config["exec_stop"] is not None:
-            exec_stop.append(self.config["exec_stop"])
-        exec_poststop = self._stop_network() + exec_poststop
-        if self.config["exec_poststop"] is not None:
-            exec_poststop.append(self.config["exec_poststop"])
+        event = Event(self, scope=event_scope)
 
-        if self.config["jail_zfs"] is True:
-            share_storage = libioc.ZFSShareStorage.QueuingZFSShareStorage(
-                jail=self,
-                logger=self.logger
-            )
-            share_storage.umount_zfs_shares()
-            exec_stop += share_storage.read_commands("jail")
-            exec_poststop += share_storage.read_commands()
+        yield event.begin()
 
-        if self.running and (os.path.isfile(self.script_env_path) is False):
-            # when a jail was started from other iocage variants
-            self._write_temporary_script_env()
-            exec_poststop.append(f"rm \"{shlex.quote(self.script_env_path)}\"")
+        config_value = self.config[f"exec_{hook}"]
 
-        self._write_hook_script(
-            "prestop",
-            self._wrap_hook_script_command_string(exec_prestop)
+        if config_value is None:
+            yield event.skip()
+            return
+
+        stdout, stderr, code = exec_func(
+            ["/bin/sh", "-c", config_value],
+            **exec_args
         )
-        self._write_hook_script(
-            "stop",
-            self._wrap_hook_script_command_string(
-                exec_stop,
-                jailed=True,
-                ignore_errors=True
-            )
-        )
-        self._write_hook_script(
-            "poststop",
-            self._wrap_hook_script_command_string(
-                exec_poststop,
-                write_env=False,
-                ignore_errors=True
-            )
-        )
+        event.stdout = stdout
+        event.stderr = stderr
+        event.code = code
+        if code > 0:
+            if force is True:
+                yield event.skip("failed")
+            else:
+                yield event.fail(f"exited with {code}")
+                raise libioc.errors.JailHookFailed(
+                    jail=self,
+                    hook="prestop",
+                    logger=self.logger
+                )
+        else:
+            yield event.end()
 
     def stop(
         self,
@@ -926,31 +924,18 @@ class JailGenerator(JailResource):
             self.require_jail_running(log_errors=log_errors)
 
         events: typing.Any = libioc.events
-        jailDestroyEvent = events.JailDestroy(self, scope=event_scope)
+        jailStopEvent = events.JailStop(self, scope=event_scope)
 
-        self._prepare_stop()
+        yield jailStopEvent.begin()
 
-        yield jailDestroyEvent.begin()
-        try:
-            self._write_jail_conf(force=force)
-            self._destroy_jail(log_errors=log_errors)
-        except Exception as e:
-            if force is True:
-                yield jailDestroyEvent.skip()
-                self.logger.debug(
-                    "Manually executing prestop and poststop hooks"
-                )
-                try:
-                    for hook_name in ["prestop", "poststop"]:
-                        libioc.helpers.exec(
-                            command=[self.get_hook_script_path(hook_name)]
-                        )
-                except Exception as e:
-                    self.logger.warn(str(e))
-            else:
-                yield jailDestroyEvent.fail(e)
-                raise e
-        yield jailDestroyEvent.end()
+        yield from self.__run_hook("prestop", force, jailStopEvent.scope)
+        yield from self.__run_hook("stop", force, jailStopEvent.scope)
+        yield from self.__destroy_jail(jailStopEvent.scope)
+        yield from self.__run_hook("poststop", force, jailStopEvent.scope)
+        yield from self.__teardown_mounts(force, jailStopEvent.scope)
+        yield from self.__clear_resource_limits(force, jailStopEvent.scope)
+
+        yield jailStopEvent.end()
 
     def _write_temporary_script_env(self) -> None:
         self.logger.debug(
@@ -1509,28 +1494,36 @@ class JailGenerator(JailResource):
             ["/usr/bin/login"] + self.config["login_flags"]
         )
 
-    def _destroy_jail(self, log_errors: bool=True) -> None:
+    def __destroy_jail(
+        self,
+        event_scope: typing.Optional['libioc.events.Scope']=None
+    ) -> typing.Generator['libioc.events.JailStop', None, None]:
 
-        stdout, stderr, returncode = self._exec_host_command(
-            [
-                "/usr/sbin/jail",
-                "-v",
-                "-r",
-                "-f",
-                self._jail_conf_file,
-                self.identifier
-            ],
-            passthru=False,
-            env=self.env
+        jailRemoveEvent = libioc.events.JailRemove(
+            jail=self,
+            scope=event_scope
         )
 
-        self.query_jid()
+        yield jailRemoveEvent.begin()
 
-        if returncode > 0:
-            raise libioc.errors.JailDestructionFailed(
-                jail=self,
-                logger=(self.logger if log_errors else None)
-            )
+        if self.running is False:
+            yield jailRemoveEvent.skip()
+            return
+
+        jid = self.jid
+        try:
+            libjail.dll.jail_remove(self.jid)
+            self.query_jid()
+            yield jailRemoveEvent.end()
+            return
+        except Exception:
+            pass
+
+        yield jailRemoveEvent.fail(f"libc jail_remove failed for {jid}")
+        raise libioc.errors.JailDestructionFailed(
+            jail=self,
+            logger=self.logger
+        )
 
     @property
     def _dhcp_enabled(self) -> bool:
@@ -1930,11 +1923,42 @@ class JailGenerator(JailResource):
             ]))
         return commands
 
-    def _clear_resource_limits(self) -> typing.List[str]:
-        if self.config['rlimits'] is False:
-            return []
+    def __clear_resource_limits(
+        self,
+        force: bool,
+        event_scope: typing.Optional['libioc.events.Scope']=None
+    ) -> typing.Generator['libioc.events.JailResourceLimitAction', None, None]:
+        jailResourceLimitActionEvent = libioc.events.JailResourceLimitAction(
+            jail=self,
+            scope=event_scope
+        )
+
+        yield jailResourceLimitActionEvent.begin()
+
+        config_value = self.config['rlimits']
+        if (config_value is False) or (config_value is None):
+            yield jailResourceLimitActionEvent.skip()
+            return
+
         self.logger.verbose("Clearing resource limits")
-        return [f"/usr/bin/rctl -r jail:{self.identifier} 2>/dev/null || true"]
+        _, _, returncode = libioc.helpers.exec(
+            [
+                "/usr/bin/rctl",
+                "-r",
+                f"jail:{self.identifier}"
+            ],
+            ignore_error=True
+        )
+
+        if returncode > 0:
+            yield jailResourceLimitActionEvent.fail()
+            if force is False:
+                raise libioc.errors.ResourceLimitActionFailed(
+                    action=f"clear resource limits of jail {self.full_name}",
+                    logger=self.logger
+                )
+
+        yield jailResourceLimitActionEvent.end()
 
     @property
     def _allow_mount(self) -> int:
@@ -2039,9 +2063,18 @@ class JailGenerator(JailResource):
                 logger=(self.logger if log_errors else None)
             )
 
-    def _teardown_mounts(self) -> typing.List[str]:
+    def __teardown_mounts(
+        self,
+        force: bool,
+        event_scope: typing.Optional['libioc.events.Scope']=None,
+    ) -> typing.Generator[libioc.events.TeardownJailMounts, None, None]:
 
-        commands: typing.List[str] = []
+        teardownJailMountsEvent = libioc.events.TeardownJailMounts(
+            jail=self,
+            scope=event_scope
+        )
+
+        yield teardownJailMountsEvent.begin()
 
         fstab_destinations = [
             line["destination"] for line in self.fstab
@@ -2065,28 +2098,57 @@ class JailGenerator(JailResource):
         ))
 
         mountpoints = fstab_destinations + system_mountpoints
+        error = False
 
-        commands.append(" ".join(libioc.helpers.umount_command(
-            mountpoints,
-            force=True,
+        _, _, returncode = libioc.helpers.exec(
+            libioc.helpers.umount_command(
+                mountpoints,
+                force=True
+            ),
+            logger=self.logger,
             ignore_error=True
-        )))
+        )
+        error |= (returncode > 0)
 
-        commands.append(" ".join(libioc.helpers.umount_command(
-            ["-a", "-F", self.fstab.path],
-            force=True,
+        _, _, returncode = libioc.helpers.exec(
+            libioc.helpers.umount_command(
+                ["-a", "-F", self.fstab.path],
+                force=True
+            ),
+            logger=self.logger,
             ignore_error=True
-        )))
+        )
+        error |= (returncode > 0)
 
-        if self.config.legacy is True:
-            commands.append(" | ".join([
-                "mount -t nullfs",
-                "sed -r 's/(.+) on (.+) \\(nullfs, .+\\)$/\\2/'",
-                f"grep '^{self.root_dataset.mountpoint}/'",
-                "xargs umount"
-            ]))
+        if (self.config.legacy is True) and error:
+            try:
+                libioc.helpers.exec(
+                    [
+                        "/bin/sh", "-c",
+                        " | ".join([
+                            "mount -t nullfs",
+                            "sed -r 's/(.+) on (.+) \\(nullfs, .+\\)$/\\2/'",
+                            f"grep '^{self.root_dataset.mountpoint}/'",
+                            "xargs umount"
+                        ])
+                    ],
+                    logger=self.logger
+                )
+            except Exception:
+                error = True
+        else:
+            error = False
 
-        return commands
+        if error is True:
+            yield teardownJailMountsEvent.fail()
+            if force is False:
+                name = self.full_name
+                raise libioc.errors.UnmountFailed(
+                    reason=f"failed to unmount mountpoints of jail {name}",
+                    logger=self.logger
+                )
+
+        yield teardownJailMountsEvent.end()
 
     def _get_absolute_path_from_jail_asset(
         self,

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -1489,8 +1489,11 @@ class JailGenerator(JailResource):
 
         jid = self.jid
         try:
-            libjail.dll.jail_remove(self.jid)
-            self.query_jid()
+            libjail.dll.jail_remove(jid)
+            while libjail.is_jid_dying(jid) is True:
+                # wait for death
+                continue
+            self.__jid = None
             yield jailRemoveEvent.end()
             return
         except Exception:

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -31,6 +31,7 @@ import shutil
 
 import libzfs
 import freebsd_sysctl
+import jail as libjail
 
 import libioc.Types
 import libioc.errors
@@ -2061,47 +2062,17 @@ class JailGenerator(JailResource):
             )
         ))
 
-        mountpoints = fstab_destinations + system_mountpoints
         error = False
-
-        _, _, returncode = libioc.helpers.exec(
-            libioc.helpers.umount_command(
-                mountpoints,
-                force=True
-            ),
-            logger=self.logger,
-            ignore_error=True
-        )
-        error |= (returncode > 0)
-
-        _, _, returncode = libioc.helpers.exec(
-            libioc.helpers.umount_command(
-                ["-a", "-F", self.fstab.path],
-                force=True
-            ),
-            logger=self.logger,
-            ignore_error=True
-        )
-        error |= (returncode > 0)
-
-        if (self.config.legacy is True) and error:
-            try:
-                libioc.helpers.exec(
-                    [
-                        "/bin/sh", "-c",
-                        " | ".join([
-                            "mount -t nullfs",
-                            "sed -r 's/(.+) on (.+) \\(nullfs, .+\\)$/\\2/'",
-                            f"grep '^{self.root_dataset.mountpoint}/'",
-                            "xargs umount"
-                        ])
-                    ],
-                    logger=self.logger
-                )
-            except Exception:
+        for mountpoint in (fstab_destinations + system_mountpoints):
+            if os.path.ismount(mountpoint) is False:
+                continue
+            if libjail.dll.unmount(mountpoint.encode("utf-8")) == 0:
+                self.logger.verbose(f"{mountpoint} successfully unmounted")
+            else:
                 error = True
-        else:
-            error = False
+                self.logger.warn(f"{mountpoint} unmount failed")
+                if force is False:
+                    break
 
         if error is True:
             yield teardownJailMountsEvent.fail()

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -886,7 +886,7 @@ class JailGenerator(JailResource):
                 yield event.fail(f"exited with {code}")
                 raise libioc.errors.JailHookFailed(
                     jail=self,
-                    hook="prestop",
+                    hook=hook,
                     logger=self.logger
                 )
         else:

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -565,6 +565,7 @@ class JailGenerator(JailResource):
                 stdout, stderr, returncode = self._launch_persistent_jail(
                     passthru=passthru
                 )
+                self.query_jid()
             else:
                 stdout, stderr, returncode = self._launch_single_command_jail(
                     single_command,

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -674,10 +674,6 @@ class JailGenerator(JailResource):
             started_jails=started_jails
         )
 
-    def _run_poststop_hook_manually(self) -> None:
-        self.logger.debug("Running poststop hook manually")
-        libioc.helpers.exec(self.get_hook_script_path("poststop"))
-
     def _wrap_jail_command(
         self,
         commands: typing.Optional[typing.List[str]]
@@ -945,35 +941,6 @@ class JailGenerator(JailResource):
         self._ensure_script_dir()
         with open(self.script_env_path, "w") as f:
             f.write(f"export IOC_JID={self.jid}")
-
-    def _write_jail_conf(self, force: bool=False) -> None:
-        if force is True:
-            stop_command = "/usr/bin/true"
-        else:
-            stop_command = (
-                f"[ -f \"{self._relative_hook_script_dir}/stop.sh\" ]"
-                " || exit 0; "
-                f". {self._relative_hook_script_dir}/stop.sh"
-            )
-
-        content = "\n".join([
-            self.identifier + " {",
-            (
-                "exec.prestop = "
-                f"\"/bin/sh {self.get_hook_script_path('prestop')}\";"
-            ), (
-                "exec.poststop = "
-                f"\"/bin/sh {self.get_hook_script_path('poststop')}\";"
-            ), (
-                f"exec.stop = \"{stop_command}\";"
-            ), (
-                f"exec.jail_user = {self._get_value('exec_jail_user')};"
-            ),
-            "}"
-        ])
-        self.logger.debug(f"Writing jail.conf file to {self._jail_conf_file}")
-        with open(self._jail_conf_file, "w") as f:
-            f.write(content)
 
     @property
     def _jail_conf_file(self) -> str:

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -1611,8 +1611,6 @@ class JailGenerator(JailResource):
             f"exec.timeout={self._get_value('exec_timeout')}",
             f"stop.timeout={self._get_value('stop_timeout')}",
             f"exec.prestart=\"{self.get_hook_script_path('prestart')}\"",
-            f"exec.prestop=\"{self.get_hook_script_path('prestop')}\"",
-            f"exec.poststop=\"{self.get_hook_script_path('poststop')}\"",
             f"exec.jail_user={self._get_value('exec_jail_user')}",
             f"mount.fstab={self.fstab.path}",
             f"mount.devfs={self._get_value('mount_devfs')}",
@@ -1704,8 +1702,7 @@ class JailGenerator(JailResource):
                     f"/usr/sbin/jexec {self.identifier} "
                     f"{self._relative_hook_script_dir}/command.sh"
                     " 2>&1"
-                ),
-                f"/bin/sh {self.get_hook_script_path('poststop')}"
+                )
             ]
         ))
 
@@ -1780,7 +1777,7 @@ class JailGenerator(JailResource):
     def _write_hook_script(self, hook_name: str, command_string: str) -> None:
         file = self.get_hook_script_path(hook_name)
         existed = os.path.isfile(file)
-        if hook_name in ["created", "poststart", "prestop"]:
+        if hook_name in ["created", "poststart"]:
             _identifier = str(shlex.quote(self.identifier))
             _jls_command = f"/usr/sbin/jls -j {_identifier} jid"
             command_string = (
@@ -1788,12 +1785,6 @@ class JailGenerator(JailResource):
                 f"$({_jls_command} 2>&1 || echo -1)"
                 "\n" + command_string
             )
-        if hook_name == "poststop":
-            command_string = (
-                "[ -f \"$(dirname $0)/.env\" ] && "
-                ". \"$(dirname $0)/.env\""
-                "\n"
-            ) + command_string
         with open(file, "w") as f:
             f.write("\n".join([
                 "#!/bin/sh",

--- a/libioc/Jail.py
+++ b/libioc/Jail.py
@@ -570,7 +570,12 @@ class JailGenerator(JailResource):
                     single_command,
                     passthru=passthru
                 )
-            self.query_jid()
+                scope = jailLaunchEvent.scope
+                yield from self.__destroy_jail(scope)
+                yield from self.__run_hook("poststop", False, scope)
+                yield from self.__teardown_mounts(False, scope)
+                yield from self.__clear_resource_limits(False, scope)
+
             if returncode != 0:
                 raise libioc.errors.JailLaunchFailed(
                     jail=self,

--- a/libioc/Network.py
+++ b/libioc/Network.py
@@ -70,9 +70,12 @@ class Network:
     ) -> None:
 
         self.logger = libioc.helpers_object.init_logger(self, logger)
-        self.firewall = libioc.Firewall.QueuingFirewall(
+        self.queuing_firewall = libioc.Firewall.QueuingFirewall(
             logger=self.logger,
             insecure=True
+        )
+        self.firewall = libioc.Firewall.Firewall(
+            logger=self.logger
         )
 
         if nic is not None:
@@ -101,55 +104,49 @@ class Network:
             self.__require_bridge()
         return self.__create_vnet_iface()
 
-    def teardown(self) -> typing.List[str]:
+    def teardown(self, jid: typing.Optional[int]=None) -> None:
         """
         Teardown the applied changes.
 
         After Jails are stopped the devices that were used by it remain on the
         host. This method is called by jails after they terminated.
         """
-        commands: typing.List[str] = []
-
         if self.vnet is False:
-            return commands
+            return
 
-        commands += self.__down_host_interface()
+        if jid is None:
+            jid = self.jail.jid
+
+        self.__down_host_interface(jid)
 
         if self._is_secure_vnet_bridge is True:
-            commands += self.__down_secure_mode_devices()
-            self.firewall.delete_rule("$IOC_JID")
-            commands += self.firewall.read_commands()
-
-        return commands
+            self.__down_secure_mode_devices(jid)
+            self.firewall.delete_rule(jid)
 
     def __require_bridge(self) -> None:
         if (self.bridge is None):
             raise libioc.errors.VnetBridgeMissing(logger=self.logger)
 
-    def __down_host_interface(self) -> typing.List[str]:
-        nic = libioc.NetworkInterface.QueuingNetworkInterface(
-            name=f"{self._escaped_nic_name}:$IOC_JID",
+    def __down_host_interface(self, jid: int) -> None:
+        libioc.NetworkInterface.NetworkInterface(
+            name=f"{self._escaped_nic_name}:{jid}",
             extra_settings=["destroy"],
             logger=self.logger,
             insecure=True
         )
-        commands: typing.List[str] = nic.read_commands()
-        return commands
 
-    def __down_secure_mode_devices(self) -> typing.List[str]:
+    def __down_secure_mode_devices(self, jid: int) -> None:
         self.logger.verbose("Downing secure mode devices")
-        commands: typing.List[str] = []
         secure_mode_nics = [
-            f"{self._escaped_nic_name}:$IOC_JID:a",
-            f"{self._escaped_nic_name}:$IOC_JID:net"
+            f"{self._escaped_nic_name}:{jid}:a",
+            f"{self._escaped_nic_name}:{jid}:net"
         ]
         for nic in secure_mode_nics:
-            commands += libioc.NetworkInterface.QueuingNetworkInterface(
+            libioc.NetworkInterface.NetworkInterface(
                 name=nic,
                 extra_settings=["destroy"],
                 insecure=True
-            ).read_commands()
-        return commands
+            )
 
     @property
     def mtu(self) -> int:
@@ -271,7 +268,7 @@ class Network:
             raise libioc.errors.VnetBridgeMissing(logger=self.logger)
 
         if self._is_secure_vnet_bridge is True:
-            self.firewall.ensure_firewall_enabled()
+            self.queuing_firewall.ensure_firewall_enabled()
 
         commands_created += self.__create_new_epair_interface(
             variable_name_a=f"IOC_NIC_EPAIR_A_{self._nic_hash}",
@@ -400,7 +397,7 @@ class Network:
                         f"Firewall permit not possible for address '{address}'"
                     )
                     continue
-                self.firewall.add_rule(firewall_rule_number, [
+                self.queuing_firewall.add_rule(firewall_rule_number, [
                     "allow", protocol,
                     "from", _address, "to", "any",
                     "layer2",
@@ -408,7 +405,7 @@ class Network:
                     "via", f"{self._escaped_nic_name}:$IOC_JID:b",
                     "out"
                 ], insecure=True)
-                self.firewall.add_rule(firewall_rule_number, [
+                self.queuing_firewall.add_rule(firewall_rule_number, [
                     "allow", protocol,
                     "from", "any", "to", _address,
                     "layer2",
@@ -416,27 +413,27 @@ class Network:
                     "via", f"{self._escaped_nic_name}:$IOC_JID",
                     "out"
                 ], insecure=True)
-                self.firewall.add_rule(firewall_rule_number, [
+                self.queuing_firewall.add_rule(firewall_rule_number, [
                     "allow", protocol,
                     "from", "any", "to", _address,
                     "via", f"{self._escaped_nic_name}:$IOC_JID",
                     "out"
                 ], insecure=True)
-            self.firewall.add_rule(firewall_rule_number, [
+            self.queuing_firewall.add_rule(firewall_rule_number, [
                 "deny", "log", protocol,
                 "from", "any", "to", "any",
                 "layer2",
                 "via", f"{self._escaped_nic_name}:$IOC_JID:b",
                 "out"
             ], insecure=True)
-            self.firewall.add_rule(firewall_rule_number, [
+            self.queuing_firewall.add_rule(firewall_rule_number, [
                 "deny", "log", protocol,
                 "from", "any", "to", "any",
                 "via", f"{self._escaped_nic_name}:$IOC_JID",
                 "out"
             ], insecure=True)
         self.logger.debug("Firewall rules added")
-        commands: typing.List[str] = self.firewall.read_commands()
+        commands: typing.List[str] = self.queuing_firewall.read_commands()
         return commands
 
     def __up_host_if(self) -> typing.List[str]:

--- a/libioc/ZFSShareStorage.py
+++ b/libioc/ZFSShareStorage.py
@@ -51,33 +51,6 @@ class ZFSShareStorage:
         self.logger.verbose("Mounting ZFS shares")
         self._mount_jail_datasets(auto_create=auto_create)
 
-    def umount_zfs_shares(self) -> None:
-        """Unmount a running jails shared ZFS datasets."""
-        for dataset in self.get_zfs_datasets():
-            self.logger.verbose(f"Unmounting ZFS Dataset {dataset.name}")
-            self._exec([
-                "/sbin/zfs",
-                "set",
-                "jailed=off",
-                shlex.quote(dataset.name)
-            ])
-            self._exec([
-                "/sbin/zfs",
-                "umount",
-                shlex.quote(dataset.name)
-            ])
-            self._exec([
-                "/sbin/zfs",
-                "mount",
-                shlex.quote(dataset.name)
-            ])
-            self._exec([
-                "/sbin/zfs",
-                "set",
-                "jailed=on",
-                shlex.quote(dataset.name)
-            ])
-
     def get_zfs_datasets(
         self,
         auto_create: bool=False

--- a/libioc/errors.py
+++ b/libioc/errors.py
@@ -211,6 +211,20 @@ class JailNotTemplate(JailException):
         JailException.__init__(self, message=msg, jail=jail, logger=logger)
 
 
+class JailHookFailed(JailException):
+    """Raised when the jail could not be launched."""
+
+    def __init__(
+        self,
+        jail: 'libioc.Jail.JailGenerator',
+        hook: str,
+        logger: typing.Optional['libioc.Logger.Logger']=None
+    ) -> None:
+        self.hook = hook
+        msg = f"Jail {jail.full_name} hook {hook} failed"
+        JailException.__init__(self, message=msg, jail=jail, logger=logger)
+
+
 class JailLaunchFailed(JailException):
     """Raised when the jail could not be launched."""
 
@@ -458,6 +472,18 @@ class ResourceLimitUnknown(IocException, KeyError):
         logger: typing.Optional['libioc.Logger.Logger']=None
     ) -> None:
         msg = f"The specified resource limit is unknown"
+        IocException.__init__(self, message=msg, logger=logger)
+
+
+class ResourceLimitActionFailed(IocException, KeyError):
+    """Raised when a resource limit has is unknown."""
+
+    def __init__(
+        self,
+        action: str,
+        logger: typing.Optional['libioc.Logger.Logger']=None
+    ) -> None:
+        msg = f"Resource Limit failed to {action}"
         IocException.__init__(self, message=msg, logger=logger)
 
 
@@ -733,12 +759,15 @@ class UnmountFailed(IocException):
 
     def __init__(
         self,
-        mountpoint: typing.Any,
+        mountpoint: typing.Any=None,
         reason: typing.Optional[str]=None,
         logger: typing.Optional['libioc.Logger.Logger']=None
     ) -> None:
 
-        msg = f"Unmount of {mountpoint} failed"
+        if mountpoint is None:
+            msg = "Umount failed"
+        else:
+            msg = f"Unmount of {mountpoint} failed"
         if reason is not None:
             msg += f": {reason}"
         super().__init__(message=msg, logger=logger)

--- a/libioc/events.py
+++ b/libioc/events.py
@@ -343,17 +343,52 @@ class JailRename(JailEvent):
         JailEvent.__init__(self, jail=jail, message=message, scope=scope)
 
 
-class JailDestroy(JailEvent):
+class JailStop(JailEvent):
     """Destroy the jail."""
 
-    def __init__(
-        self,
-        jail: 'libioc.Jail.JailGenerator',
-        message: typing.Optional[str]=None,
-        scope: typing.Optional[Scope]=None
-    ) -> None:
+    pass
 
-        JailEvent.__init__(self, jail=jail, message=message, scope=scope)
+
+class JailRemove(JailEvent):
+    """Remove the jail(2)."""
+
+    pass
+
+
+class TeardownJailMounts(JailStop):
+    """Teardown a jails mountpoints."""
+
+    pass
+
+
+class JailResourceLimitAction(JailEvent):
+    """Set or unset a jails resource limits."""
+
+    pass
+
+
+class JailHook(JailEvent):
+    """Run jail hook."""
+
+    pass
+
+
+class JailHookPrestop(JailHook):
+    """Run jail prestop hook."""
+
+    pass
+
+
+class JailHookStop(JailHook):
+    """Run jail prestop hook."""
+
+    pass
+
+
+class JailHookPoststop(JailHook):
+    """Run jail prestop hook."""
+
+    pass
 
 
 class JailFstabUpdate(JailEvent):

--- a/libioc/events.py
+++ b/libioc/events.py
@@ -361,6 +361,12 @@ class TeardownJailMounts(JailStop):
     pass
 
 
+class TeardownJailNetwork(JailStop):
+    """Teardown a jails mountpoints."""
+
+    pass
+
+
 class JailResourceLimitAction(JailEvent):
     """Set or unset a jails resource limits."""
 
@@ -925,6 +931,7 @@ class JailCommandExecution(JailEvent):
         """Successfully finish an event."""
         self.stdout = stdout
         return IocEvent.end(self, message)
+
 
 # PKG
 

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,4 @@
 ucl==0.8.0
 gitpython
 freebsd_sysctl==0.0.5
-jail==0.0.4
+jail==0.0.6

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -199,7 +199,8 @@ def new_jail(
         zfs=zfs
     )
     yield new_jail
-    new_jail.destroy(force=True)
+    new_jail.stop(force=True)
+    new_jail.destroy()
 
 
 @pytest.fixture(scope="function")
@@ -208,7 +209,7 @@ def existing_jail(
     local_release: 'libioc.Release.ReleaseGenerator',
 ) -> 'libioc.Jail.Jail':
     new_jail.create(local_release)
-    return new_jail
+    yield new_jail
 
 
 @pytest.fixture(scope="function")

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -203,6 +203,15 @@ def new_jail(
 
 
 @pytest.fixture(scope="function")
+def existing_jail(
+    new_jail: 'libioc.Jail.Jail',
+    local_release: 'libioc.Release.ReleaseGenerator',
+) -> 'libioc.Jail.Jail':
+    new_jail.create(local_release)
+    return new_jail
+
+
+@pytest.fixture(scope="function")
 def bridge_interface() -> str:
     bridge_name = "bridgeTest" + str(random.randint(1024, 4096))
     subprocess.check_output(

--- a/tests/test_Jail.py
+++ b/tests/test_Jail.py
@@ -23,8 +23,10 @@
 # IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 """Unit tests for Jail."""
+import typing
 import json
 import os
+import subprocess
 
 import pytest
 import libzfs
@@ -69,9 +71,60 @@ class TestJail(object):
         assert data["basejail"] != 1
         assert data["basejail"] != True
 
+    def test_can_be_started(
+        self,
+        existing_jail: 'libioc.Jail.Jail'
+    ) -> None:
+        """Test if a jail can be started."""
+        existing_jail.config["mount_devfs"] = False
+        existing_jail.save()
+
+        assert existing_jail.running is False
+        existing_jail.start()
+        assert existing_jail.running is True
+
+        stdout = subprocess.check_output(
+            [f"/sbin/mount | grep {existing_jail.root_dataset.name}"],
+            shell=True
+        ).decode("utf-8")
+
+        assert "launch-scripts in stdout"
+        assert stdout.strip().count("\n") == 1
+
+    def test_can_be_stopped(
+        self,
+        existing_jail: 'libioc.Jail.Jail'
+    ) -> None:
+        """Test if a jail can be started."""
+        existing_jail.start()
+        assert existing_jail.running is True
+        existing_jail.stop()
+        assert existing_jail.running is False
+
 
 class TestNullFSBasejail(object):
     """Run tests for NullFS Basejails."""
+
+    @pytest.fixture
+    def basedirs(self, host: 'libioc.Host.Host'):
+        basedirs = [
+            "bin",
+            "boot",
+            "lib",
+            "libexec",
+            "rescue",
+            "sbin",
+            "usr/bin",
+            "usr/include",
+            "usr/lib",
+            "usr/libexec",
+            "usr/sbin",
+            "usr/share",
+            "usr/libdata",
+        ]
+        if host.distribution.name == "FreeBSD":
+            basedirs.append("usr/lib32")
+        return basedirs
 
     def test_can_be_created(
         self,
@@ -99,3 +152,46 @@ class TestNullFSBasejail(object):
         assert data["basejail"] == "yes"
         if "basejail_type" in data:
             assert data["basejail_type"] == "nullfs"
+
+    def test_can_be_started(
+        self,
+        existing_jail: 'libioc.Jail.Jail',
+        basedirs: typing.List[str]
+    ) -> None:
+        """Test if a jail can be started."""
+        existing_jail.config["mount_devfs"] = False
+        existing_jail.config["basejail"] = True
+        existing_jail.save()
+
+        assert existing_jail.running is False
+        existing_jail.start()
+        assert existing_jail.running is True
+
+        root_path = existing_jail.root_dataset.name
+        stdout = subprocess.check_output(
+            [f"/sbin/mount | grep {existing_jail.root_dataset.name}"],
+            shell=True
+        ).decode("utf-8")
+        assert "launch-scripts in stdout"
+        assert stdout.strip().count("\n") == 1 + len(basedirs)
+        for basedir in basedirs:
+            assert f"{root_path}/{basedir}" in stdout
+
+    def test_can_be_stopped(
+        self,
+        existing_jail: 'libioc.Jail.Jail',
+        basedirs: typing.List[str]
+    ) -> None:
+        """Test if a jail can be started."""
+        existing_jail.start()
+        assert existing_jail.running is True
+        existing_jail.stop()
+        assert existing_jail.running is False
+
+        root_path = existing_jail.root_dataset.name
+        stdout = subprocess.check_output(
+            [f"/sbin/mount | grep {root_path}"],
+            shell=True
+        ).decode("utf-8")
+        for basedir in basedirs:
+            assert f"{root_path}/{basedir}" not in stdout

--- a/tests/test_Jail.py
+++ b/tests/test_Jail.py
@@ -188,6 +188,11 @@ class TestNullFSBasejail(object):
         existing_jail.stop()
         assert existing_jail.running is False
 
+        with pytest.raises(subprocess.CalledProcessError):
+            stdout = subprocess.check_output(
+                [f"/usr/sbin/jls", "-j", existing_jail.identifier]
+            ).decode("utf-8")
+
         root_path = existing_jail.root_dataset.name
         stdout = subprocess.check_output(
             [f"/sbin/mount | grep {root_path}"],

--- a/tests/test_VNET.py
+++ b/tests/test_VNET.py
@@ -37,20 +37,19 @@ class TestVNET(object):
 
     def test_vnet_without_interfaces_can_only_see_lo0(
         self,
-        new_jail: 'libioc.Jail.Jail',
-        local_release: 'libioc.Release.ReleaseGenerator'
+        existing_jail: 'libioc.Jail.Jail'
     ) -> None:
         """Test if a VNET jail without interfaces can only see lo0."""
-        new_jail.config["vnet"] = True
-        new_jail.create(local_release)
+        existing_jail.config["vnet"] = True
+        existing_jail.save()
 
-        assert len(list(new_jail.networks)) == 0
+        assert len(list(existing_jail.networks)) == 0
 
-        new_jail.start()
+        existing_jail.start()
 
         stdout_lines = subprocess.check_output(
-            ["/usr/sbin/jexec", str(new_jail.jid), "/sbin/ifconfig"]
-        ).decode().split("\n")
+            ["/usr/sbin/jexec", str(existing_jail.jid), "/sbin/ifconfig"]
+        ).decode("utf-8").split("\n")
 
         # filter lines that begin with a whitespace
         configured_nics = [
@@ -64,23 +63,25 @@ class TestVNET(object):
 
     def test_static_ip_configuration(
         self,
-        new_jail: 'libioc.Jail.Jail',
-        local_release: 'libioc.Release.ReleaseGenerator',
+        existing_jail: 'libioc.Jail.Jail',
         bridge_interface: str
     ) -> None:
         """Test if static IPv4 and IPv6 addresses can be configured."""
-        new_jail.config["vnet"] = True
-        new_jail.config["interfaces"] = f"vnet23:{bridge_interface}"
-        new_jail.config["ip4_addr"] = f"vnet23|172.16.99.23/24"
-        new_jail.create(local_release)
+        existing_jail.config["vnet"] = True
+        existing_jail.config["interfaces"] = f"vnet23:{bridge_interface}"
+        existing_jail.config["ip4_addr"] = f"vnet23|172.16.99.23/24"
+        existing_jail.save()
 
-        assert len(list(new_jail.networks)) == 1
+        assert len(list(existing_jail.networks)) == 1
 
-        new_jail.start()
+        existing_jail.start()
 
-        stdout = subprocess.check_output(
-            ["/usr/sbin/jexec", str(new_jail.jid), "/sbin/ifconfig", "vnet23"]
-        ).decode()
+        stdout = subprocess.check_output([
+            "/usr/sbin/jexec",
+            str(existing_jail.jid),
+            "/sbin/ifconfig",
+            "vnet23"
+        ]).decode("utf-8")
 
         # filter lines that begin with a whitespace
         assert "172.16.99.23" in stdout

--- a/tests/test_VNET.py
+++ b/tests/test_VNET.py
@@ -85,3 +85,30 @@ class TestVNET(object):
 
         # filter lines that begin with a whitespace
         assert "172.16.99.23" in stdout
+
+    def test_vnet_interfaces_are_removed_on_stop(
+        self,
+        existing_jail: 'libioc.Jail.Jail',
+        bridge_interface: str
+    ) -> None:
+        """Test if VNET interfaces are removed from the host on stop."""
+        existing_jail.config["vnet"] = True
+        existing_jail.config["interfaces"] = f"vnet22:{bridge_interface}"
+        existing_jail.config["ip4_addr"] = f"vnet22|172.16.99.23/24"
+        existing_jail.save()
+
+        existing_jail.start()
+        jid = existing_jail.jid
+
+        stdout = subprocess.check_output(
+            ["/sbin/ifconfig", f"vnet22:{jid}"]
+        ).decode("utf-8")
+
+        assert "UP" in stdout
+
+        existing_jail.stop()
+
+        with pytest.raises(subprocess.CalledProcessError) as excinfo:
+            stdout = subprocess.check_output(
+                ["/sbin/ifconfig", f"vnet22:{jid}"]
+            ).decode("utf-8")

